src/libcore/slice/rotate.rs

```diff
@@ -2,88 +2,170 @@ use crate::cmp;
+/// ^   ^   ^    ^    ^       ^    ^    ^
 /// ```
+/// Fortunately, the number of skipped over elements between finalized elements is always equal, so
+/// we can just offset our starting position and do more rounds (the total number of rounds is the
+/// `gcd(left + right, right)` value). The end result is that all elements are finalized once and
+/// only once.
+///
+/// Algorithm 2 is used if `left + right` is large but `min(left, right)` is small enough to
+/// fit onto a stack buffer. The `min(left, right)` elements are copied onto the buffer, `memmove`
+/// is applied to the others, and the ones on the buffer are moved back into the hole on the
+/// opposite side of where they originated.
+///
+/// Algorithms that can be vectorized outperform the above once `left + right` becomes large enough.
+/// Algorithm 1 can be vectorized by chunking and performing many rounds at once, but there are too
+/// few rounds on average until `left + right` is enormous, and the worst case of a single
+/// round is always there. Instead, algorithm 3 utilizes repeated swapping of
+/// `min(left, right)` elements until a smaller rotate problem is left.
 ///
-/// Once the rotation is small enough, copy some elements into a stack
-/// buffer, `memmove` the others, and move the ones back from the buffer.
-pub unsafe fn ptr_rotate<T>(mut left: usize, mid: *mut T, mut right: usize) {
+/// ```text
+/// left = 11, right = 4
+/// [4 5 6 7 8 9 10 11 12 13 14 . 0 1 2 3]
+///                  ^  ^  ^  ^   ^ ^ ^ ^ swapping the right most elements with elements to the left
+/// [4 5 6 7 8 9 10 . 0 1 2 3] 11 12 13 14
+///        ^ ^ ^  ^   ^ ^ ^ ^ swapping these
+/// [4 5 6 . 0 1 2 3] 7 8 9 10 11 12 13 14
+/// we cannot swap any more, but a smaller rotation problem is left to solve
+/// ```
+/// when `left < right` the swapping happens from the left instead.
+pub unsafe fn ptr_rotate<T>(mut left: usize, mut mid: *mut T, mut right: usize) {
+    type BufType = [usize; 32];
+    if mem::size_of::<T>() == 0 {
+        return;
+    }
     loop {
-        let delta = cmp::min(left, right);
-        if delta <= RawArray::<T>::capacity() {
-            // We will always hit this immediately for ZST.
-            break;
+        // N.B. the below algorithms can fail if these cases are not checked
+        if (right == 0) || (left == 0) {
+            return;
         }
-
-        ptr::swap_nonoverlapping(
-            mid.sub(left),
-            mid.add(right - delta),
-            delta);
-
-        if left <= right {
-            right -= delta;
+        if left + right < 24 {
+            // Algorithm 1
+            // Microbenchmarks indicate that the average performance for random shifts is better all
+            // the way until about `left + right == 32`, but the worst case performance breaks even
+            // around 16. 24 was chosen as middle ground.
+            // Additionally, algorithm 1 is still faster than the others when `T` is huge.
```

Comment:
This seems like an interesting observation; should it be reflected in the code somehow?  Like add `|| mem::size_of::<T>() >= 64` to the condition here or something?  (As a rough guess that if something is about the size of a cache line or more then random reads aren't a problem the way they are with spread-out byte reads.)

Reply:
I just made a breakthrough. Before, I always assumed that algorithm 1 would never work beyond a slice of size 32 regardless of `T` size. I think I tried out a `||` modifier before, but I was trying to use `mem::size_of::<BufType>()`. If instead I use the `BufType` for algorithm 2 exclusively and use a different size for the algorithm 1 branch modifier, I end up with great results.
On my machine at least, adding a `|| (mem::size_of::<T>() > mem::size_of::<[usize; 4]>())` significantly improves performance all the way up, except for [usize; 7]. Prime slice sizes don't seem to degrade this performance significantly like it does for the single usize benches.

Reply:
Sounds great!